### ISSUE TITLE
3 kinds of errors on a fresh test: (1) 'NoneType' object cannot be interpreted as an integer (2) 'NoneType' object is not subscriptable (3) Download appears to be stuck

### ISSUE THREAD

**Issue by holta** [iiab/calibre-web#167]
Thanks @nzola for having confirmed!

> I could only copy 2 videos. Most of them failed. Single videos and playlist. I re-installed multipass and Calibre-Web on my new computer.

![IMG-20240602-WA0000~4](https://github.com/iiab/calibre-web/assets/2458907/433af3d1-3c7a-407f-a211-8b97b17e2faa)

**Comment by holta** [comment]
Possibly related:

- #114
- #135
- #158

**Comment by holta** [comment]
@nzola for clarififies that these were some of the videos he tried above:

- https://www.youtube.com/watch?v=zOHi37jWcro (25 Advanced English Vocabulary Phrases)
- https://www.youtube.com/watch?v=xQpImggreLE (March 19, 2024, ASCODERU)
- https://www.youtube.com/@iogacademy/playlists (IOG Academy)
- https://www.youtube.com/watch?v=5WenPx1krGE&t=90s (Essential Cardano360 May 2024 Edition)
- https://www.youtube.com/@IohkIo/playlists (Input Output)

**Comment by nzola** [comment]
Last night, I successfully upgraded Calibre-Web on a month year old PI4 and run the video playlist "https://www.youtube.com/@IohkIo/playlists", this is the results I got on the task view and status view:
![image](https://github.com/iiab/calibre-web/assets/17815091/f84ac278-6c56-4e8b-9460-ec8c3024b213)
![image](https://github.com/iiab/calibre-web/assets/17815091/8e1d1b42-85d3-4975-a3f2-7a1273a9953d)

**Comment by holta** [comment]
@nzola

You've pasted in the wrong iiab-diagnostics:

- https://dpaste.com/BGKGBZK7G shows a PC CPU, rather than a Raspberry Pi CPU.

- Please run **iiab-diagnostics** on the Raspberry Pi 4 and post the resulting URL to this ticket when you can, Thanks!

**Comment by nzola** [comment]
> @nzola
> 
> You've pasted in the wrong iiab-diagnostics:
> 
> * https://dpaste.com/BGKGBZK7G shows a PC CPU, rather than a Raspberry Pi CPU.
> * Please run **iiab-diagnostics** on the Raspberry Pi 4 and post the resulting URL (a dpaste.com pastebin) to this ticket when you can, Thanks!
PUBLISHING TO URL...
https://dpaste.com/ELHD529SS

**Comment by holta** [comment]
> * (1) **'NoneType' object cannot be interpreted as an integer** : likely due to a missing metadata field like view_count or duration during the metadata fetching task. Restricted videos will throw this error.

Thanks @deldesir and @EMG70 and @nzola:

Browsing to https://youtu.be/ROgfEuRI6Sw shows:

> If the owner of this video has granted you access, please **sign in**.

While seemingly obvious:

- Videos are hidden from and disappear from YouTube regularly / eternally for a variety of different reasons, consequently...
- **IIAB Teachers expect and deserve IIAB Calibre-Web "Tasks" view to show a "Googlable" _(or very friendly!!)_ message explaining exactly what they should try next~**

🎨

**Comment by deldesir** [comment]
You're right. The "NoneType" objectcan stay in the log file, but the user should see a message explaining what's going on. I'll work on this.

**Comment by holta** [comment]
PS A message like "'NoneType' object cannot be interpreted as an integer" strongly indicates that Python variable type needs validation?! 🙃

- Let's gather all such reproducible cases.  As much as we can in coming weeks ✊

- And stress test these during our move to Ruff-based CI (GitHub Actions) and gradually increasing _validation of Python variable types._

**Comment by holta** [comment]
@deldesir

@nzola says that playlist https://www.youtube.com/@iogacademy/playlists still continues to fail to download — does this mean that older playlists with videos that are **"unavailable"** or **"owner of this video has [NOT] granted you access"** need more sophisticated handling?

1. On PC/VM it failed with:

> Metadata fetch: https://www.youtube.com/watch?v=V0zV_xO3Yal failed: 'NoneType' object cannot be interpreted as an integer

https://dpaste.com/CDSUG68CD

2. On RPi 4 it failed with:

> Metadata fetch: https://www.youtube.com/watch?v=ROgfEuRI6Sw failed: 'NoneType' object cannot be interpreted as an integer

https://dpaste.com/6V66CQ7NQ

**Comment by deldesir** [comment]
I see, I'm addressing this with #175 and another PR which will make calculations of views per day failsafe.

**Comment by holta** [comment]
> @nzola says that playlist https://www.youtube.com/@iogacademy/playlists still continues to fail to download — does this mean that older playlists with videos that are **"unavailable"** or **"owner of this video has [NOT] granted you access"** need more sophisticated handling?

@deldesir it appears playlists/channels are blocked from downloading if even a single YouTube video is **"unavailable"** or **"owner of this video has [NOT] granted you access"**.

- Can this strict approach be made more pragmatic — for educators wanting the bulk of the playlist/channel downloaded — even despite a few videos no longer being available?

- _(Instead of blocking the full download!)_

**Comment by holta** [comment]
> > @nzola says that playlist https://www.youtube.com/@iogacademy/playlists still continues to fail to download — does this mean that older playlists with videos that are **"unavailable"** or **"owner of this video has [NOT] granted you access"** need more sophisticated handling?
> 
> @deldesir it appears playlists/channels are blocked from downloading if even a single YouTube video is **"unavailable"** or **"owner of this video has [NOT] granted you access"**.
> 
> * Can this strict approach be made more pragmatic — for educators wanting the bulk of the playlist/channel downloaded — even despite a few videos no longer being available?
> * _(Instead of blocking the full download!)_

@deldesir's PR #179 proposes a solution to this problem, please see:

- https://github.com/iiab/calibre-web/pull/179#issuecomment-2160579297
